apm-agent-plugins/apm-scala-concurrent-plugin/src/main/java/co/elastic/apm/agent/scala/concurrent/FutureInstrumentation.java

```diff
@@ -0,0 +1,101 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.scala.concurrent;
+
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation;
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import scala.concurrent.Promise;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.*;
+
+public abstract class FutureInstrumentation extends ElasticApmInstrumentation {
+
+    @VisibleForAdvice
+    @SuppressWarnings("WeakerAccess")
+    public static final WeakConcurrentMap<Promise<?>, TraceContextHolder<?>> promisesToContext =
+        new WeakConcurrentMap.WithInlinedExpunction<>();
+
+    @Nonnull
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("concurrent", "future");
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("scala.concurrent.impl.Promise$Transformation"));
+    }
+
+    public static class ConstructorInstrumentation extends FutureInstrumentation {
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return isConstructor();
```

Comment:
Promises could be created outside a transaction but used in the context of a transaction at the point a promise is scheduled. 
Is there a canonical place where promises are scheduled? Maybe something equivalent to `Executor#execute(Runnable)`?

Reply:
Mmmm, Ill have a look. 

Do you mean in the context of, someone creates a `Promise` outside of a transaction, without evaluating it. And then only start evaluating it later on, inside of a transaction?

Reply:
Yes, exactly.

This would also allow us to do another nice thing. Currenly, all classes have to be evaluated whether they are a subclass of `Promise` (`hasSuperType(named("scala.concurrent.impl.Promise$Transformation"))`). This is a rather costly matcher as we can't exclude any class by name so all classes that are loaded have to be parsed.

See this example of the AsyncHttpClient instrumentation how we only instrument `AsyncHandler` classes that are passed to the `org.asynchttpclient.DefaultAsyncHttpClient#executeRequest(org.asynchttpclient.Request, org.asynchttpclient.AsyncHandler<T>)` method.

https://github.com/elastic/apm-agent-java/blob/75160feaf475a467ec347b32a98b24767d70cc52/apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AbstractAsyncHttpClientInstrumentation.java#L106

Using this strategy, we wouldn't have to instrument any `Promise`s to start with. Only those subclasses that pass the canonical place where they are scheduled would be instrumented at runtime.

Reply:
But how would you handle the case that there is a thread-switch between the creation of a Promise and the scheduling? 
If the context is not propagated across that thread-switch, you would not know the transaction linked to this Promise and schedule it linked to no transaction.

Reply:
Usually, the thread on which the promise is scheduled is the thread we want to propagate the context. 
Do you know any counter-examples?

Reply:
1. You start a transaction (thread 1)
2. You create a Promise (thread 1) - no context is saved on this Promise now
3. Some thread switching - as I understand, the transaction of thread 1 will not be propagated
4. Promise gets scheduled (thread 2) - it cannot find an active transaction

Reply:
> Some thread switching 

This thread switching should also be covered by one of our instrumentation. For example, if the switch is done by another promise.

Reply:
Oke, I'll dig a bit deeper into how Scala executes the Promises.

Reply:
I've added some instrumentation on Scala's ExecutionContext instead of the Promise directly.
Some parts are working but Im having issues with `Future.sequence` which transforms a list of Future to one Future. 

For this, it uses the following logic:
```scala
/**
   * WARNING: Only ever execute logic which will quickly return control to the caller.
   *
   * This `ExecutionContext` steals execution time from other threads by having its
   * `Runnable`s run on the `Thread` which calls `execute` and then yielding back control
   * to the caller after *all* its `Runnable`s have been executed.
   * Nested invocations of `execute` will be trampolined to prevent uncontrolled stack space growth.
   *
   * When using `parasitic` with abstractions such as `Future` it will in many cases be non-deterministic
   * as to which `Thread` will be executing the logic, as it depends on when/if that `Future` is completed.
   *
   * Do *not* call any blocking code in the `Runnable`s submitted to this `ExecutionContext`
   * as it will prevent progress by other enqueued `Runnable`s and the calling `Thread`.
   * 
   * Symptoms of misuse of this `ExecutionContext` include, but are not limited to, deadlocks
   * and severe performance problems.
   *
   * Any `NonFatal` or `InterruptedException`s will be reported to the `defaultReporter`.
   */
  object parasitic extends ExecutionContextExecutor with BatchingExecutor {
    override final def submitForExecution(runnable: Runnable): Unit = runnable.run()
    override final def execute(runnable: Runnable): Unit = submitSyncBatched(runnable)
    override final def reportFailure(t: Throwable): Unit = defaultReporter(t)
  }
```

During debugging, I see the `SpanInScopeRunnableWrapper` be correctly submitted to this parasitic executor. But after the execution of the `Future.sequence`, the context seems to be lost.

Reply:
Not sure if relevant, but in DEBUG mode, I see the following:
```
2020-04-24 20:14:59.428 [ForkJoinPool-1-worker-19] DEBUG co.elastic.apm.agent.mdc.MdcActivationListener - Class loader jdk.internal.loader.ClassLoaders$AppClassLoader@512ddf17 cannot load log4j API
java.lang.ClassNotFoundException: org.apache.log4j.MDC
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at co.elastic.apm.agent.mdc.MdcActivationListener$2.get(MdcActivationListener.java:76)
	at co.elastic.apm.agent.mdc.MdcActivationListener$2.get(MdcActivationListener.java:70)
	at co.elastic.apm.agent.cache.WeakKeySoftValueLoadingCache$CacheValue.<init>(WeakKeySoftValueLoadingCache.java:124)
	at co.elastic.apm.agent.cache.WeakKeySoftValueLoadingCache$CacheValue.<init>(WeakKeySoftValueLoadingCache.java:116)
	at co.elastic.apm.agent.cache.WeakKeySoftValueLoadingCache.get(WeakKeySoftValueLoadingCache.java:81)
	at co.elastic.apm.agent.mdc.MdcActivationListener.beforeActivate(MdcActivationListener.java:154)
	at co.elastic.apm.agent.impl.transaction.TraceContextHolder.activate(TraceContextHolder.java:101)
	at co.elastic.apm.agent.impl.transaction.AbstractSpan.activate(AbstractSpan.java:345)
	at co.elastic.apm.agent.impl.transaction.AbstractSpan.activate(AbstractSpan.java:42)
	at co.elastic.apm.api.AbstractSpanImpl.activate(AbstractSpanImpl.java:137)
	at me.milan.MainF$.$anonfun$program$3(MainF.scala:30)
```

Reply:
@felixbarny Should I create an issue about the mismatch between traces in the UI and the actual logging of the transaction-id?

Reply:
> One of the issues that I have is the logging though. With %X{transaction.id:-null} in my logback, I had all transaction ids of every operation when instrumenting the Promises directly.

Hm, that's strange.. Maybe there's another thread switch in-between? Could there be some work-stealing going on? Is it seemingly random when the MDC works and when it doesn't? In the cases where there's no MDC, does `ElasticApm.currentSpan()` (from the `apm-agent-api` module) return a noop span?


The debug message about `ClassNotFoundException: org.apache.log4j.MDC` can be ignored.

Reply:
At the moment it looks quite random. I'll into into it a bit more to see if I can reproduce it in a better way.
But as far as I understand it, the logging should be part of the submitted `Runnable` and therefore have exactly the same transaction as the action it is logging.

Reply:
Yes, that's correct. Whenever a span is activated, it also sets the MDC.

Reply:
> In the cases where there's no MDC, does ElasticApm.currentSpan() (from the apm-agent-api module) return a noop span?

Would you test this by adding doing something like:
```
println(ElasticApm.currentSpan())
log.info(testing MDC)
```

Reply:
Yes, exactly

Reply:
@felixbarny The `println` of `ElasticApm.currentSpan().getId()` returns an empty String so according to the docs it is indeed a noop.

I'll need to check how the logging is executed in a different thread for whatever reason.

Reply:
It seems to be related to the usage of the Scala sttp library. It wraps the `AsyncHttpClient` from Java with https://github.com/reactive-streams/reactive-streams-jvm.

Reply:
@felixbarny So basically what happens is that the Scala future will be completed (scheduled?) the moment `override def onCompleted(): Unit` is called on the `org.asynchttpclient.handler.StreamedAsyncHandler`.

If I add an `ElasticAPM.currentSpan()` in the `onCompleted`, it prints a Noop which shows that no trace-context is available at that moment. And therefore, also not on the Scala Future that gets completed by that method.

I assume because the `AsyncHttpClient` is instrumented on the `executeRequest`, it causes the span of the http request to still be correctly closed. 

But because I continue logging after the completion of the Scala future, the logging lost its context due to the `org.asynchttpclient.handler.StreamedAsyncHandler`.

Reply:
> So basically what happens is that the Scala future will be completed (scheduled?) the moment override def onCompleted(): Unit is called on the org.asynchttpclient.handler.StreamedAsyncHandler

Thanks for the analysis!

> Unless you see it as a mistake that the trace-context is not passed into the org.asynchttpclient.handler.StreamedAsyncHandler#onCompleted?

Yes, that's a mistake.

This should be easy to fix in the AsyncHttpClient instrumentation:

https://github.com/elastic/apm-agent-java/blob/94c92159d99f57d67d6f2ca0aa16e373082f1684/apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AbstractAsyncHttpClientInstrumentation.java#L174-L187

We should activate the span in the `@OnMethodEnter` advice and deactivate it in the `@OnMethodExit` advice.

Reply:
@felixbarny Cool :) Can I try to create a PR to fix the AsyncHttpClient or do you prefer to do it from your side?

Once that is merged, Ill continue with testing this PR and start cleaning it up.

Reply:
> Can I try to create a PR to fix the AsyncHttpClient or do you prefer to do it from your side?

Sure, go ahead! I think we should do the activation on all currently instrumented `AsyncHandler` callback methods. This way, users can add custom stuff in their own callbacks to the spans like `ElasticApm.currentSpan().addCustom("foo", responseStatus.getStatusText());`

Reply:
@felixbarny First attempt: https://github.com/elastic/apm-agent-java/pull/1172
Will need some help on how to add proper tests for this though.

Reply:
@felixbarny I missed a method `onStream` on #1172 so will create a follow PR to add that one as well.
Locally, it seems the span is now correctly passed along the `AsyncHandler` but there is still a place in my test setup where the span is not correctly passed on through all the Scala Futures.

I will investigate further but I am maybe thinking there was a reason why other solutions instrument the `Promise` directly and not the `ExecutionContext`.

Reply:
Both approaches seem to have their pros and cons. I currently don't have a clear opinion which is the best approach. I'm leaning towards instrument on `ExecutionContext#execute` as it's more efficient (no startup overhead) and supports propagating context for `Promise`s that are created outside of a transaction. But I also see that it requires more work and is potentially more brittle.

Reply:
I went into the code again and found the behavior described in the accepted answer on https://stackoverflow.com/questions/59642314/scala-future-with-callback-and-context-switching.

What seems to happen is that if a `Future` is already completed, it will be executed immediately. Meaning that `final AbstractSpan<?> active = getActive();` in the `OnExecute` instrumentation, may return a null since the callback is executed on a thread that does not have an active span on it.

Since the `scala.concurrent.impl.Promise.DefaultPromise#dispatchOrAddCallback` is the method that is executing the `Future`, this method somehow needs to be able to track which span is linked to that `Future/Promise`.

I think this forces the choice to instrument on the `Future/Promise` directly instead of with the `ExecutionContext`.

One if the issues with instrumenting on the `Promise` directly, seems to be the extra dependency on `Scala` in the agent which cause some strange `LinkageErrors` sometimes. Is this related to the reason why the Elastic agent needs to shade libraries to avoid certain conflicts?

Reply:
I added the instrumentation on the Promise back. It all works now for my test-setup.

I sometimes get a warning though:
`09:35:28.230 [ForkJoinPool-5-worker-19] WARN co.elastic.apm.agent.impl.ElasticApmTracer - Tried to start a span with a non-zero reference count 1 'GET postman-echo.com' 00-00e98d3ae2d1d3665c8286557ca405f1-05f18334e53a3ee6-01 (3c295061)`

As well as the LinkageError in some cases:
```
[ERROR] Exception in thread "main" java.lang.LinkageError: loader constraint violation: loader 'bootstrap' wants to load class scala.collection.immutable.LazyList$State$Empty$. A different class with the same name was previously loaded by 'app'. (scala.collection.immutable.LazyList$State$Empty$ is in unnamed module of loader 'app')
```

Reply:
@felixbarny As discussed, I make the Map use an `Object` instead of a `Promise` but it still get the Linkage issue. Here is the full trace:

```
scala-elastic-apm[ERROR] Exception in thread "main" java.lang.LinkageError: loader constraint violation: loader 'bootstrap' wants to load class scala.collection.immutable.LazyList$State$Empty$. A different class with the same name was previously loaded by 'app'. (scala.collection.immutable.LazyList$State$Empty$ is in unnamed module of loader 'app')
scala-elastic-apm[ERROR] 	at scala.collection.immutable.LazyList$.<clinit>(LazyList.scala:949)
scala-elastic-apm[ERROR] 	at scala.package$.<clinit>(package.scala:89)
scala-elastic-apm 2020-05-15 13:09:22.792 [elastic-apm-init-instrumentation-shutdown-hook] INFO co.elastic.apm.agent.impl.ElasticApmTracer - Tracer switched to STOPPED state
scala-elastic-apm[ERROR] 	at scala.Predef$.<clinit>(Predef.scala:154)
scala-elastic-apm 2020-05-15 13:09:22.792 [elastic-apm-init-instrumentation-shutdown-hook] INFO co.elastic.apm.agent.report.ApmServerReporter - dropped events because of full queue: 0
scala-elastic-apm[ERROR] 	at scala.concurrent.Future$.<clinit>(Future.scala:520)
scala-elastic-apm[ERROR] 	at me.milan.main.future.Main$.delayedEndpoint$me$milan$main$future$Main$1(Main.scala:32)
scala-elastic-apm[ERROR] 	at me.milan.main.future.Main$delayedInit$body.apply(Main.scala:19)
scala-elastic-apm[ERROR] 	at scala.Function0.apply$mcV$sp(Function0.scala:39)
scala-elastic-apm[ERROR] 	at scala.Function0.apply$mcV$sp$(Function0.scala:39)
scala-elastic-apm[ERROR] 	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
scala-elastic-apm[ERROR] 	at scala.App.$anonfun$main$1(App.scala:73)
scala-elastic-apm[ERROR] 	at scala.App.$anonfun$main$1$adapted(App.scala:73)
scala-elastic-apm[ERROR] 	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
scala-elastic-apm[ERROR] 	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
scala-elastic-apm[ERROR] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:920)
scala-elastic-apm[ERROR] 	at scala.App.main(App.scala:73)
scala-elastic-apm[ERROR] 	at scala.App.main$(App.scala:71)
scala-elastic-apm[ERROR] 	at me.milan.main.future.Main$.main(Main.scala:19)
```

Reply:
Actually, the usage of `Promise` as a type parameter of the map shouldn't be a problem because of type erasure. 

What you could try is to use `Object` instead of `Promise<?>` in the advice method signatures and cast within the method body. Maybe that avoids some eager linking. For example:

```java
        @Advice.OnMethodExit(suppress = Throwable.class)
        public static void onExit(@Advice.This Object thiz) {
            final AbstractSpan<?> context = getActive();
            if (context != null) {
                promisesToContext.put(thiz, context);
            }
        }
```

Reply:
Some things:
* When I use `Object` in the Map and method so that I can remove the Scala dependency, there is no LinkageError.
* Using Promise in the Map and in the method parameters Object does still have the LinkageError
* The LinkageError does only happen when forking the JVM (in case of running Debug from IntelliJ or using something like https://github.com/spray/sbt-revolver to start your application)

Reply:
The LinkageError also happens when Cinnamon is enabled (and without running a forked JVM). So it seems specific to the Scala dependencies clashing somehow.

On the WARN that is happening:
`scala-elastic-apm 2020-05-15 14:08:53.201 [ForkJoinPool-1-worker-23] WARN co.elastic.apm.agent.impl.ElasticApmTracer - Tried to start a span with a non-zero reference count 1 'SELECT' 00-fbed3a699f6c244dffa8de3a3d95e48d-bd4655c8959cd4c8-01 (7b9e3d51)`

Im confused on how this warning can be generated. The method that logs this warning, first creates a new Span, and then counts the references. I don't see how the new Span can suddenly have any references right after being made and before the method actually returns the created Span.

Reply:
I've added suggestions that are likely to fix the cause of this.

> I don't see how the new Span can suddenly have any references right after being made and before the method actually returns the created Span.

That's because the span it gets out of the pool is still being referenced.

Reply:
The WARN is gone indeed :) 

So ElasticAPM uses a pool of Spans that it cycles through? And since the Span is creating in the Promise but only activated later on, it may be that the Span gets recycled before it is actually used?

Incrementing and decrementing makes sure the Span does not get recycled?

Reply:
Exactly. The pool is basically a queue that is initially empty. When trying to get a span from an empty pool, a new span is created. If the reference count if a span is 0, it gets recycled. Activating a span increments the reference count so under normal circumstances, you don’t have to worry about the ref count. But in those async scenarios, it’s important to keep the span alive if it can be used after it’s ended.

Reply:
> And since the Span is creating in the Promise but only activated later on, it may be that the Span gets recycled before it is actually used?

Actually, it's not creating any span for the Promise. It just propagates the span that has been active when the promise was created to the thread that executes the promise. But the span might be already ended once the promise is executed.